### ISSUE TITLE
Training and reasoning code

### ISSUE THREAD

**Issue by 0317lwj** [streamer-AP/CrowdLocater#1]
Thank you very much for the help you have provided. But I can't use the weights you provided directly. I was wondering if you could provide the full training and inference code. If so I will reference your work in my work. Once again, thanks a lot.

**Comment by streamer-AP** [comment]
I have provided a complete inference code and it should work if you set the data in the right place. Please let me know what problem you are encountering.

**Comment by Edward-Cousins** [comment]
I Was unable to access the weights provided. Could you provide link for weights in google drive instead of Baidu.

**Comment by 0317lwj** [comment]
where is “build_model” 
from models import build_model
https://github.com/streamer-AP/CrowdLocater/blob/8135bd3ca1d20215ce48025cc87bb67223ec91f3/inference.py#L18

**Comment by 0317lwj** [comment]
Hi, I noticed that the weighting parameters you provided do not match the structure of the HRNet-W48 network at the last layer, can you provide me with your network configuration file

**Comment by streamer-AP** [comment]
Add a __init__.py under folder "model"
from .counting_model import build_counting_model
def build_model(args):
    if args.name == "counting":
        return build_counting_model(args)

**Comment by streamer-AP** [comment]
And here is the model structure
import torch
import torch.nn as nn
import torch.nn.functional as F
from .utils import MaxFiltering


class FFN(nn.Module):
    def __init__(self, in_chans, mid_chans, out_chans):
        super().__init__()
        self.conv1 = nn.Conv2d(in_chans, mid_chans, 1)
        self.act1 = nn.ReLU()
        self.conv2 = nn.Conv2d(mid_chans, out_chans, 1)
        self.act2 = nn.ReLU()

    def forward(self, x):
        x = self.act1(self.conv1(x))
        x = self.act2(self.conv2(x))
        return x


class fidt(nn.Module):
    def __init__(self):
        super(fidt, self).__init__()
        self.conv1 =nn.Sequential(
            nn.Conv2d(96,96,1),
            nn.BatchNorm2d(96),
            nn.ReLU(),
        )
        self.conv2 = nn.Sequential(
            nn.Conv2d(192, 96, 1),
            nn.BatchNorm2d(96),
            nn.ReLU(),
        )
        self.conv3 = nn.Sequential(
            nn.Conv2d(384, 96, 1),
            nn.BatchNorm2d(96),
            nn.ReLU(),
        )

        self.out1 = nn.Sequential(            
            nn.Conv2d(96+96+96, 256, 1),
            nn.BatchNorm2d(256),
            nn.ReLU(),
            nn.Conv2d(256, 128, 3,dilation=2,padding=2),
            nn.ReLU(),
            nn.Conv2d(128, 64, 3,dilation=2,padding=2),
            nn.ReLU(),
            nn.Conv2d(64, 1, 1),
            nn.ReLU(),
        )
        self._initialize_weights()

    def _initialize_weights(self):
        for m in self.modules():
            if isinstance(m, nn.Conv2d):
                nn.init.normal_(m.weight, std=0.01)
                if m.bias is not None:
                    nn.init.constant_(m.bias, 0)
            elif isinstance(m, nn.BatchNorm2d):
                nn.init.constant_(m.weight, 1)
                nn.init.constant_(m.bias, 0)

    def forward(self, x):
        # print(x.shape)
        # torch.Size([1, 96, 256, 256])
        # torch.Size([1, 192, 128, 128])
        # torch.Size([1, 384, 64, 64])
        # torch.Size([1, 768, 32, 32])
        x1, x2 , x3  = x
        x1 = self.conv1(x1)
        x2 = self.conv2(x2)
        x3 = self.conv3(x3)
        # x3 = self.conv3(x3)
        # x4 = self.conv4(x4)

        
        x3=  F.interpolate(x3, scale_factor=4)
        x2 = F.interpolate(x2, scale_factor=2)

        z=torch.cat([x1,x2,x3],dim=1)

        out1 = self.out1(z)
        out1 = F.interpolate(out1, scale_factor=2)
        out_dict = {}
        out_dict["predict_counting_map"] = out1
        return out_dict


def build_counting_head(args):

    return fidt()

**Comment by streamer-AP** [comment]
The model structure
```python

import torch
import torch.nn as nn
import torch.nn.functional as F
from .utils import MaxFiltering


class FFN(nn.Module):
    def __init__(self, in_chans, mid_chans, out_chans):
        super().__init__()
        self.conv1 = nn.Conv2d(in_chans, mid_chans, 1)
        self.act1 = nn.ReLU()
        self.conv2 = nn.Conv2d(mid_chans, out_chans, 1)
        self.act2 = nn.ReLU()

    def forward(self, x):
        x = self.act1(self.conv1(x))
        x = self.act2(self.conv2(x))
        return x


class fidt(nn.Module):
    def __init__(self):
        super(fidt, self).__init__()
        self.conv1 =nn.Sequential(
            nn.Conv2d(96,96,1),
            nn.BatchNorm2d(96),
            nn.ReLU(),
        )
        self.conv2 = nn.Sequential(
            nn.Conv2d(192, 96, 1),
            nn.BatchNorm2d(96),
            nn.ReLU(),
        )
        self.conv3 = nn.Sequential(
            nn.Conv2d(384, 96, 1),
            nn.BatchNorm2d(96),
            nn.ReLU(),
        )

        self.out1 = nn.Sequential(            
            nn.Conv2d(96+96+96, 256, 1),
            nn.BatchNorm2d(256),
            nn.ReLU(),
            nn.Conv2d(256, 128, 3,dilation=2,padding=2),
            nn.ReLU(),
            nn.Conv2d(128, 64, 3,dilation=2,padding=2),
            nn.ReLU(),
            nn.Conv2d(64, 1, 1),
            nn.ReLU(),
        )
        self._initialize_weights()

    def _initialize_weights(self):
        for m in self.modules():
            if isinstance(m, nn.Conv2d):
                nn.init.normal_(m.weight, std=0.01)
                if m.bias is not None:
                    nn.init.constant_(m.bias, 0)
            elif isinstance(m, nn.BatchNorm2d):
                nn.init.constant_(m.weight, 1)
                nn.init.constant_(m.bias, 0)

    def forward(self, x):
        # print(x.shape)
        # torch.Size([1, 96, 256, 256])
        # torch.Size([1, 192, 128, 128])
        # torch.Size([1, 384, 64, 64])
        # torch.Size([1, 768, 32, 32])
        x1, x2 , x3  = x
        x1 = self.conv1(x1)
        x2 = self.conv2(x2)
        x3 = self.conv3(x3)
        # x3 = self.conv3(x3)
        # x4 = self.conv4(x4)

        
        x3=  F.interpolate(x3, scale_factor=4)
        x2 = F.interpolate(x2, scale_factor=2)

        z=torch.cat([x1,x2,x3],dim=1)

        out1 = self.out1(z)
        out1 = F.interpolate(out1, scale_factor=2)
        out_dict = {}
        out_dict["predict_counting_map"] = out1
        return out_dict


def build_counting_head(args):

    return fidt()
```

**Comment by 0317lwj** [comment]
thanks, I'm not sure where the build_counting_model is for the code you provided? Nor do I know where the forward_points method is? Can you provide the complete inference code that adapts your trained weights? I saw in your paper that you mentioned that the locater was trained independently, can you tell me how you trained it, because the locater I trained didn't achieve your performance on the sensecrowd and HT21 datasets.

Translated with DeepL.com (free version)
https://github.com/streamer-AP/CrowdLocater/blob/8135bd3ca1d20215ce48025cc87bb67223ec91f3/inference.py#L93C51-L93C65

**Comment by 0317lwj** [comment]
thanks, I'm not sure where the build_counting_model is for the code you provided? Nor do I know where the forward_points method is? Can you provide the complete inference code that adapts your trained weights? I saw in your paper that you mentioned that the locater was trained independently, can you tell me how you trained it, because the locater I trained didn't achieve your performance on the sensecrowd and HT21 datasets.
https://github.com/streamer-AP/CrowdLocater/blob/8135bd3ca1d20215ce48025cc87bb67223ec91f3/inference.py#L93

**Comment by streamer-AP** [comment]
Previously, the files in the secondary directory were lost due to transfer problems. Thanks for helping me find that. Now, all files are uploaded. As for training, the only thing to note is the size of the image, as SenseCrowd and ht21 are simpler than popular datasets of Image crowd counting, so most algorithms can achieve similar results.
